### PR TITLE
Add Support for Optional Bean Properties.

### PR DESCRIPTION
**Pull request by karthicks**
For details on the problem description, please refer to [issue-399](https://github.com/typesafehub/config/issues/399).

### CODE REVIEW

**Comment by viktorklang** [comment]
See README.md and the Overview section: "implemented in plain Java with no dependencies"

File: config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java [hunk 1]
```diff
@@ -105,8 +109,18 @@
                 Method setter = beanProp.getWriteMethod();
                 Type parameterType = setter.getGenericParameterTypes()[0];
                 Class<?> parameterClass = setter.getParameterTypes()[0];
-                Object unwrapped = getValue(clazz, parameterType, parameterClass, config, originalNames.get(beanProp.getName()));
-                setter.invoke(bean, unwrapped);
+                String configPropName = originalNames.get(beanProp.getName());
+                if (configPropName == null) {
+                    configPropName = beanProp.getName();
+                }
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java]: previously the code would have thrown NullPointerException if beanProp.getName wasn't in originalNames it looks like; is this case (`configPropName == null`) actually possible? Do the unit tests find a way to trigger this codepath? (there is a coverage target in the sbt build I believe to check)

File: config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java [hunk 2]
```diff
@@ -252,4 +266,27 @@ private static boolean hasAtLeastOneBeanProperty(Class<?> clazz) {
 
         return false;
     }
+
+    private static boolean isOptionalProperty(Class beanClass, PropertyDescriptor beanProp) {
+        Field field = getField(beanClass, beanProp.getName());
+        if (field == null) {
+            throw new ConfigException.BadBean("Bean property '" + beanProp + "' of class " + beanClass.getName() + " does not exist");
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java]: this this exception possible? How did we get beanProp if there's no associated field?

**Comment by havocp** [comment]
a couple questions but looks generally good to me. @viktorklang @2m do you have any comments?

File: config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java (hunk 1 above)
Comment (karthicks) [inline config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java]: The `beanProp.getName()` will not appear in `originNames` if the `config` (file) does not contain an entry for that property to begin with. Added a test case called `testNotAnOptionalProperty` to cover this codepath.

File: config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java (hunk 2 above)
Comment (karthicks) [inline config/src/main/java/com/typesafe/config/impl/ConfigBeanImpl.java]: You're right. The check is redundant and I'll get rid of it.

**Comment by havocp** [comment]
Thanks! Merged, if people have more comments we can always merge a new PR ;-)

**Comment by karthicks** [comment]
Great, thank you! If it makes sense to use annotations to solve current or future open issues, let me know. I'll be more than happy to help.

**Pull request merged**